### PR TITLE
[Camera] Return all possible cameras on iOS

### PR DESCRIPTION
**Pull request by martingeorgiu**
The current version is returning on iOS only front and back wide-angle cameras. This PR adds telephoto and ultrawide if they are present.

Fixes https://github.com/flutter/flutter/issues/80865

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by martingeorgiu** [comment]
I didn't add any tests because this is not testable in any practical way as it just changes in calling of the native ios camera api.

**Review by hellohuanlin — APPROVED** [review]

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Per discussion in Discord, this is testable with a mock discovery service. And since this doesn't meet any of Flutter policy criteria for test exemption, it needs a test.

File: packages/camera/camera/example/ios/RunnerTests/AvailableCamerasTest.m
Comment (hellohuanlin) [inline packages/camera/camera/example/ios/RunnerTests/AvailableCamerasTest.m:12]: can we just use OCMock for this?

File: packages/camera/camera/ios/Classes/CameraPlugin.m
Comment (hellohuanlin) [inline packages/camera/camera/ios/Classes/CameraPlugin.m:110]: can this be just `discoverySessionWIthDeviceTypes: dicscoveryDevices`

File: packages/camera/camera/example/ios/RunnerTests/AvailableCamerasTest.m
Comment (martingeorgiu) [reply packages/camera/camera/example/ios/RunnerTests/AvailableCamerasTest.m:12]: well, initially I wanted to use OCMock for that, however, when I wrote:
```
id deviceMock = OCMClassMock([AVCaptureDevice class]);
OCMStub([deviceMock uniqueID]).andReturn(@"1");
```
I was getting: `Multiple methods named 'uniqueID' found with mismatched result, parameter type or attributes` and the same for `position`, so I've decided to go with mock class instead. (sorry if I'm making some obvious mistake, it's my first time with obj-c. :D)

File: packages/camera/camera/ios/Classes/CameraPlugin.m
Comment (martingeorgiu) [reply packages/camera/camera/ios/Classes/CameraPlugin.m]: yes you're right! I simplified that in the tests as well

File: packages/camera/camera/example/ios/RunnerTests/AvailableCamerasTest.m
Comment (stuartmorgan) [reply packages/camera/camera/example/ios/RunnerTests/AvailableCamerasTest.m:12]: You should not use `id` for the variable type. `id` is roughly the equivalent of `Object` in Dart, which means the compiler has absolutely no idea what object to check methods against.

Comment (martingeorgiu) [reply packages/camera/camera/example/ios/RunnerTests/AvailableCamerasTest.m]: ah, cool, now it makes sense. I changed it to use the `OCMClassMock` thx for the explanation

**Review by hellohuanlin — APPROVED** [review]
one typo. otherwise lgtm

**Comment by jmagman** [comment]
Also, please rebase onto https://github.com/flutter/plugins/pull/5428 and adjust pubspec versions.

**Comment by jmagman** [comment]
Also needs another review from @stuartmorgan since tests have been added.

File: packages/camera/camera/ios/Classes/CameraPlugin.m
Comment (martingeorgiu) [reply packages/camera/camera/ios/Classes/CameraPlugin.m]: fixed

File: packages/camera/camera/CHANGELOG.md
Comment (stuartmorgan) [inline packages/camera/camera/CHANGELOG.md:3]: Returns

See https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style

File: packages/camera/camera/example/ios/RunnerTests/AvailableCamerasTest.m
Comment (stuartmorgan) [inline packages/camera/camera/example/ios/RunnerTests/AvailableCamerasTest.m:42]: This is both more readable and more performant (the latter doesn't matter in a test of course, but it's a better practice in general) as:

```
NSMutableArray *requiredTypes = [@[AVCaptureDeviceTypeBuiltInWideAngleCamera, AVCaptureDeviceTypeBuiltInTelephotoCamera] mutableCopy];
```

File: packages/camera/camera/example/ios/RunnerTests/AvailableCamerasTest.m
Comment (stuartmorgan) [inline packages/camera/camera/example/ios/RunnerTests/AvailableCamerasTest.m:92]: Same.

File: packages/camera/camera/ios/Classes/CameraPlugin.m
Comment (stuartmorgan) [inline packages/camera/camera/ios/Classes/CameraPlugin.m:105]:
```
NSMutableArray *discoveryDevices = [@[ addObject:AVCaptureDeviceTypeBuiltInWideAngleCamera, addObject:AVCaptureDeviceTypeBuiltInTelephotoCamera ] mutableCopy];
```

(run through clang-format, of course.)

**Review by stuartmorgan — APPROVED** [review]
LGTM with nits.

File: packages/camera/camera/example/ios/RunnerTests/AvailableCamerasTest.m
Comment (martingeorgiu) [reply packages/camera/camera/example/ios/RunnerTests/AvailableCamerasTest.m]: sure 👍

File: packages/camera/camera/example/ios/RunnerTests/AvailableCamerasTest.m
Comment (martingeorgiu) [reply packages/camera/camera/example/ios/RunnerTests/AvailableCamerasTest.m]: 👍

File: packages/camera/camera/ios/Classes/CameraPlugin.m
Comment (martingeorgiu) [reply packages/camera/camera/ios/Classes/CameraPlugin.m]: 👍

**Comment by martingeorgiu** [comment]
Hi guys @jmagman @stuartmorgan @hellohuanlin, I'm sorry for being a bit impatient, but is there anything I need to change or it can be merged now? I'm still resolving changelog conflicts, but the PR is already approved

**Comment by hellohuanlin** [comment]
@martingeorgiu since jenn requested stuart for another round of review and stuart stamped it with some nits, you should be good to go after fixing those nits. 

To merge, please don't hit the green button. Instead use the label "waiting for tree to go green".

**Comment by stuartmorgan** [comment]
> Hi guys @jmagman @stuartmorgan @hellohuanlin, I'm sorry for being a bit impatient, but is there anything I need to change or it can be merged now?

@martingeorgiu I wasn't aware until you commented that this was ready for another look. I can look now.

> To merge, please don't hit the green button. Instead use the label "waiting for tree to go green".

@hellohuanlin Only project members can do either of these things.

**Pull request merged**
